### PR TITLE
Insert parsing logic improved

### LINKED ISSUE #72: `insert_atomic_to_complex` and `insert_struct_to_complex` do not check if the replaced agent is unspecified (i.e. atomic has `'_'` feature and structure has no atomics specified)
If we are trying to replace atomic/structure but there is no available candidate, we should rise an exception an notify the user.

### PR DESCRIPTION
**Pull request by mopichalova**
Insert methods (insert_atomic_to_struct, insert_struct_to_complex, insert_atomic_to_complex) fixed in accordance with BCSL. 
The changes include ensuring proper matching and handling cases where the replaced agent is unspecified. 
Additionally, a new utility function, is_empty, has been added to check if an agent is empty.

Close #72

### CODE REVIEW

**Comment by mopichalova** [comment]
I am unsure about raising ComplexParsingError, especially the error that is used for the error message. Can you please check that?

File: eBCSgen/Parsing/ParseBCSL.py [hunk 1]
```diff
@@ -346,19 +352,42 @@ def insert_struct_to_complex(self, struct, complex):
...
+                    ):
+                        if self.get_name(
+                            struct.children[1].children[j]
+                        ) == self.get_name(
+                            complex.children[i].children[0].children[1].children[k]
+                        ):
+                            struct_found = False
+                            break
+                    if not struct_found:
+                        break
```
Comment (xtrojak) [inline eBCSgen/Parsing/ParseBCSL.py:368]: Please explain (or, even better, add comments to docstring of this function), what is the purpose of this part. Might be useful to clarify that we are looking at atomics and why.

File: eBCSgen/Parsing/ParseBCSL.py [hunk 2]
```diff
@@ -335,9 +336,14 @@ def insert_atomic_to_struct(self, atomic, struct):
         Adds or replaces atomic subtree in struct tree.
         """
         if len(struct.children) == 2:
+            for i in range(len(struct.children[1].children)):
+                if self.get_name(atomic) == self.get_name(
+                    struct.children[1].children[i]
+                ):
+                    raise ComplexParsingError("Matching agent was not found", struct)
```
Comment (xtrojak) [inline eBCSgen/Parsing/ParseBCSL.py:343]: I would go with something like:

```suggestion
                    raise ComplexParsingError(f"Illegal nesting sequence: {atomic}:{struct}")
```

**Review by xtrojak — COMMENTED** [review]
I think it's fine, the tests are passing and seems to be working. Just please improve the docstrings, and we are good to go.

File: eBCSgen/Parsing/ParseBCSL.py (hunk 2 above)
Comment (mopichalova) [reply eBCSgen/Parsing/ParseBCSL.py]: Thank you, exceptions are edited now.

File: eBCSgen/Parsing/ParseBCSL.py (hunk 1 above)
Comment (mopichalova) [reply eBCSgen/Parsing/ParseBCSL.py]: I added comments to `insert_struct_to_complex` and expanded docstring with details for each of the function.

**Review by xtrojak — APPROVED** [review]
Thank you! I am merging this.

**Pull request merged**